p2pclient/streams.go

```diff
@@ -0,0 +1,90 @@
+package p2pclient
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+
+	ggio "github.com/gogo/protobuf/io"
+	pb "github.com/libp2p/go-libp2p-daemon/pb"
+)
+
+// NewStream initializes a new stream on one of the protocols in protos with
+// the specified peer.
+func (c *Client) NewStream(peer []byte, protos []string) (*pb.StreamInfo, error) {
+	r := ggio.NewDelimitedReader(c.control, MessageSizeMax)
+	w := ggio.NewDelimitedWriter(c.control)
+
+	req := &pb.Request{
+		Type: pb.Request_STREAM_OPEN.Enum(),
+		StreamOpen: &pb.StreamOpenRequest{
+			Peer:  peer,
+			Proto: protos,
+		},
+	}
+
+	if err := w.WriteMsg(req); err != nil {
+		return nil, err
+	}
+
+	res := &pb.Response{}
+	if err := r.ReadMsg(res); err != nil {
+		return nil, err
+	}
+
+	if err := res.GetError(); err != nil {
+		return nil, errors.New(err.GetMsg())
+	}
+
+	return res.GetStreamInfo(), nil
+}
+
+// StreamHandlerFunc is the type of callbacks executed upon receiving a new stream
+// on a given protocol.
+type StreamHandlerFunc func(*pb.StreamInfo, io.ReadWriteCloser)
+
+// NewStreamHandler establishes an inbound unix socket and starts a listener.
+// All inbound connections to the listener are delegated to the provided
+// handler.
+func (c *Client) NewStreamHandler(proto string, path string, handler StreamHandlerFunc) (io.Closer, error) {
```

Comment:
I think we should reuse the same UNIX socket for all protocols we set a handler for -- that's what the Gerbil client does and what's natural (we have mux from the stream info object).

Reply:
I'm a bit confused here. I was expecting the application registers a handler, and the daemon just knows about it. When a peer opens a stream to us for that protocol, we push a "new incoming stream" event in the control socket, along with a path where a _new unix socket_ would be created.

Reply:
Looks like the code is forgetting to tell the daemon about the new stream handler!